### ISSUE TITLE
importC: combining union and bitfield cause an ICE

### ISSUE THREAD

**Issue by tom-tan** [ldc-developers/ldc#4477]
I found this issue when I try to use an external C library.

## How to reproduce

- sample.d
```dlang
import csrc;

void main() {}
```

- csrc.c
```c
typedef union {
    struct {
        int type:8;
        int magic32;
    } data;

    int type:8;
} U;

struct S {
    U value;
    int a;
};
```

And run the following:
```console
$ ldc2 sample.d csrc.c
```

## Expected behavior
It generates a binary without errors.

## Actual behavior
It fails with an ICE with the following message:

```console
$ ldc2 sample.d csrc.c
csrc.c(12): Error: variable `csrc.S.a` overlaps previous field. This is an ICE, please file an LDC issue.
```

## Environment
```console
$ arch
x86_64
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.2 LTS
Release:        22.04
Codename:       jammy
$ ldc2 --version | head
LDC - the LLVM D compiler (1.34.0):
  based on DMD v2.104.2 and LLVM 16.0.6
  built with LDC - the LLVM D compiler (1.34.0)
  Default target: x86_64-unknown-linux-gnu
  Host CPU: znver3
...
```

**Comment by kinke** [comment]
Translated to D, to be used with `-preview=bitfields`:
```D
union U {
    struct S {
        int type:8;
        int magic32;
    }
    S data;

    int type:8;
}

pragma(msg, U.sizeof); // 4
pragma(msg, U.alignof); // 4
pragma(msg, U.S.magic32.offsetof); // 4, lol

struct S {
    U value;
    int a;
}
```

This is a frontend issue; `sizeof(U)` in C++ is 8. Please file it upstream.

**Comment by kinke** [comment]
Oh wow - `U.S.sizeof` is 8, but `U.sizeof` 4. OMG.

**Comment by tom-tan** [comment]
Done!
https://issues.dlang.org/show_bug.cgi?id=24193

**Comment by Imperatorn** [comment]
The bug report is about the discrepancy between the size of a union with a struct and bitfields in C and D. The expected size of the union in D is 8 bytes, but the actual size is 4 bytes. This is because the dmd compiler does not properly align the struct inside the union to match the C layout.

The dmd compiler source code has a function called `aggregateLayout` in the compiler/src/dmd/aggregate.d file¹, which is responsible for computing the layout of structs and unions. This function calls another function called `alignmember` in the compiler/src/dmd/target.d file¹, which determines the alignment of each member of an aggregate type. The `alignmember` function has a special case for bitfields, where it uses the `bitFieldAlignment` function to get the alignment of a bitfield¹. The `bitFieldAlignment` function returns 1 for all bitfields, regardless of their size or type¹. This means that the struct inside the union is not aligned to 4 bytes, as it would be in C, but to 1 byte, which reduces the size of the union to 4 bytes.

To fix this bug, one possible solution is to modify the `bitFieldAlignment` function to return the alignment of the underlying type of the bitfield, instead of 1. For example, if the bitfield is of type int, it should return 4. This would make the struct inside the union aligned to 4 bytes, as it is in C, and increase the size of the union to 8 bytes. Alternatively, one could add a flag to indicate whether the bitfield is part of a union or not, and use different alignment rules depending on that flag. This would allow more flexibility and control over the layout of unions with bitfields.

**Comment by Imperatorn** [comment]
https://github.com/dlang/dmd/blob/5f524486aa64a12ebc1ecc46bd3f7f1ed6c31782/compiler/src/dmd/aggregate.d#L274

**Comment by Imperatorn** [comment]
@WalterBright i tried looking at this but I currently have covid. Maybe you can take a look 🍀

**Comment by kinke** [comment]
Was fixed upstream; trickled down to LDC in #4522.
